p2p/config.go

```diff
@@ -89,47 +160,51 @@ func (c *Config) build() error {
 	return nil
 }
 
-// setupKey attempts to load the p2p private key required to start the p2p
-// servce, if a key does not exist or cannot be loaded, it creates a new key
-// using the random seed (if random seed is not set, creates new random key)
-func (c *Config) setupKey() error {
-
-	// attempt to load existing key
-	key, err := loadKey(c.DataDir)
-	if err != nil {
-		return err
-	}
-
-	// check if key set
-	if key == nil {
-		log.Trace(
-			"Generating new p2p identity",
-			"directory", c.DataDir,
+// buildProtocol verifies and applies defaults to the protocol configuration
+func (c *Config) buildProtocol() error {
+	if c.ProtocolID == "" {
+		log.Warn(
+			"ProtocolID not defined, using DefaultProtocolID",
+			"DefaultProtocolID", DefaultProtocolID,
 		)
+		c.ProtocolID = DefaultProtocolID
+	}
 
-		// generate key
-		key, err = generateKey(c.RandSeed, c.DataDir)
-		if err != nil {
-			return err
-		}
-
-	} else {
-
-		// get p2p identity from private key
-		id, err := peer.IDFromPrivateKey(key)
-		if err != nil {
-			return err
+	if c.ProtocolVersion == 0 {
+		s := strings.Split(c.ProtocolID, "/")
+		// expecting default protocol format ("/gossamer/dot/0")
+		if len(s) != 4 {
+			log.Warn(
+				"Unable to parse ProtocolID, using DefaultProtocolVersion",
+				"DefaultProtocolVersion", 0,
+			)
+		} else {
+			// get the last item in the slice ("0" in default protocol format)
+			i, err := strconv.Atoi(s[len(s)-1])
+			if err != nil {
+				log.Warn(
+					"Unable to parse ProtocolID, using DefaultProtocolVersion",
+					"DefaultProtocolVersion", 0,
+				)
+			} else {
+				c.ProtocolVersion = uint32(i)
+			}
 		}
+	}
 
-		log.Trace(
-			"Using existing p2p identity",
-			"directory", c.DataDir,
-			"id", id,
+	if c.MinSupportedVersion < c.ProtocolVersion {
+		log.Warn("MinSupportedVersion less than ProtocolVersion, using ProtocolVersion",
+			"ProtocolVersion", c.ProtocolVersion,
 		)
+		c.MinSupportedVersion = c.ProtocolVersion
 	}
 
-	// set private key
-	c.privateKey = key
+	if c.Roles == 0 {
+		log.Warn("Roles 0 is not supported, using DefaultRoles",
+			"DefaultRoles", DefaultRoles,
+		)
+		c.Roles = DefaultRoles
+	}
```

Comment:
Roles 0 is no network according to the specification. I'm assuming no need to support it in the network service.

Reply:
i think if roles is 0, we don't turn on networking at all. maybe we should be checking it before starting up the p2p service? is roles something we can configure in the config?

Reply:
Yea, probably should be checked before starting network service. Will fix.

Reply:
#601